### ISSUE TITLE
[Bitmex] API rate limit conflict among multiple accounts

### ISSUE THREAD

**Issue by declan94** [knowm/XChange#3121]
When we create multiple `BitmexExchange` with different API keys for different account, each of them should have there own API rate limit independently. 

 The current implementation, however, fails to achieve this. Because these three properties -- `rateLimit`, `rateLimitRemaining` and `rateLimitReset` are static properties of `BitmexBaseService`, which would be globally shared.

**Comment by walec51** [comment]
Yeah, API rate limiting was done in some modules in an inconsistent and mostly faulty way.
I think we should throw out any rate limiting code for now and consider it out of the scope of this library until we design a more unified mechanism of implementing this and making it an optional feature.

**Comment by makarid** [comment]
Indeed rate limit is not mandatory for this library.

**Comment by walec51** [comment]
a new proposition in this regard: https://github.com/knowm/XChange/issues/3211

**Comment by walec51** [comment]
BTW. Binance also limits the API calls to methods that do not require any account (`/api/v1/depth` for example)
I don't think that having more then one account will allow you to call `/api/v1/depth` more often on a given IP. Could some one verify this? either by emailing Binance or doing an experimental test?

**Comment by ottuzzi** [comment]
Hi,

   the question is: why are they static?
I just checked out the code, removed the static keyword, and everything seems fine (I was expecting it cannot compile anymore).
Can it be so easy to simply remove the static keyword?

*Ciao
Piero

**Comment by walec51** [comment]
If limit rating is done on a per account level then why do they block you entire IP if you surpass it?

https://www.bitmex.com/app/restAPI#Request-Rate-Limits

> If a large number of 4xx or 5xx responses are delivered in a short period of time, your IP may be banned for an hour.

If the rate limiting is bound to your IP then rate limiting should be static and the number of `Exchange` instances you create should no affect on it.

So before hacking anything I we need to know if:

- rate limits apply to a given IP?
- or to a given account?

Therefore I repeat my question: Could some one verify this? either by emailing Binance or doing an experimental test?

**Comment by ottuzzi** [comment]
Hi @walec51,

   good question: documentation is not clear.
In one place there is what you said above (so per IP ban) but just above they say 

> Requests to our REST API are rate limited to 60 requests per 1 minute. This counter refills continuously. **If you are not logged in, your ratelimit is 30/1 minute.**

Bold is mine ;)
So it looks like they have a per customer ratelimit and a per IP ban when the client is giving too much errors.
I have a client with multiple "customers"... I will enable trace and will report here :)

*Ciao
Piero

**Comment by ottuzzi** [comment]
OK, I'm pretty sure the HTTP Headers returned by bitmex are per account: I clearly can see they "restart" from 60 for each customer.

**Comment by walec51** [comment]
ok, thank you for the research, I will take it into consideration in my new rate limiting implementation  #3211

for now feel free to add a PR with the change you suggested, it should work, please test it out

**Comment by declan94** [comment]
@walec51 The thing is when the API key is not provided (only public endpoints are accessible), the rate limit is applied to the IP, while the API key is provided, it is applied to the account.

**Comment by treerz** [comment]
Hi @ottuzzi ,

i think simply removing the static keyword would just create another problem: a conflict between different service types.

My understanding is that these fields are declared as _static_ at the `BitmexBaseService` level so that they will be shared among the different service types (`BitmexAccountService`, `BitmexMarketDataService`, `BitmexTradeService`).
If you use more that one type of service, and each tracks the rate limits independently, that would lead to a similar conflict, with out-of-sync rate-limit information in each.
